plugin/pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -665,31 +698,20 @@ func (s *ServiceAffinity) CheckServiceAffinity(pod *api.Pod, meta interface{}, n
 	// check if the pod being scheduled has the affinity labels specified in its NodeSelector
 	affinityLabels := FindLabelsInSet(s.labels, labels.Set(pod.Spec.NodeSelector))
 
-	// Introspect services IFF we didn't predefine all the affinity labels in the pod itself.
+	// Step 2: If Step 1 didnt have all constraints, introspect nodes to find the missing constraints.
 	if len(s.labels) > len(affinityLabels) {
-		services, err := s.serviceLister.GetPodServices(pod)
-		if err == nil && len(services) > 0 {
-			// just use the first service and get the other pods within the service
-			// TODO: a separate predicate can be created that tries to handle all services for the pod
-			selector := labels.SelectorFromSet(services[0].Spec.Selector)
-			servicePods, err := s.podLister.List(selector)
-			if err != nil {
-				return false, nil, err
-			}
-			// consider only the pods that belong to the same namespace
-			nsServicePods := FilterPodsByNamespace(servicePods, pod.Namespace)
-			if len(nsServicePods) > 0 {
-				// consider any service pod and fetch the node its hosted on
-				otherNode, err := s.nodeInfo.GetNodeInfo(nsServicePods[0].Spec.NodeName)
+		if len(predicateMeta.serviceAffinityMatchingPodServices) > 0 {
```

Comment:
Instead of lines 688-690 here what you should do here is to:
```
var services []*api.Service
if pm, ok := meta.(*predicateMetadata); ok {
  services = pm.serviceAffinityMatchingPodServices
} else {
  var err
  services, err = s.serviceLister.GetPodServices(pod)
  if err != nil {
    glog.Errorf(...)
  }
}
```

With this approach, this predicate will work no matter if the metadata is passed or not - the difference will be that with metadata is significantly more efficient.